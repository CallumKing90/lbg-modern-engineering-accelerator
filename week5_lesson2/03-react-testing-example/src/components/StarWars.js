import React, { useEffect, useState } from 'react'
import axios from 'axios'

const StarWars = () => {
  const [films, setFilms] = useState([])

  useEffect(() => {
    const source = axios.CancelToken.source()
    axios
      .get('https://swapi.dev/api/films/', { cancelToken: source.token })
      .then(({ data }) => {
        setFilms(data.results)
      })
      .catch((error) => {
        if (axios.isCancel(error)) {
          console.log(`Request cancelled: ${error.message}`)
        } else {
          console.error(error)
        }
      })

    return () => {
      source.cancel('Cancelling axios in cleanup')
    }
  }, [])

  return (
    <main>
      <ul>
        {films.map(({ episode_id, title, opening_crawl }) => (
          <li key={episode_id}>
            <h2>{title}</h2>
            <p>{opening_crawl}</p>
          </li>
        ))}
      </ul>
    </main>
  )
}

export default StarWars
