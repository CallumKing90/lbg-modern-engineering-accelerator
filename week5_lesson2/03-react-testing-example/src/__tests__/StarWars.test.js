import React from 'react'
import { act, screen, render, waitFor } from '@testing-library/react'

// import API mocking utilities from Mock Service Worker
import { rest } from 'msw'
import { setupServer } from 'msw/node'

import StarWars from '../components/StarWars'

const server = setupServer(
  // capture "GET" requests to the Star Wars API
  rest.get('https://swapi.dev/api/films/', (_req, res, ctx) => {
    // Respond using a mocked JSON body
    return res(
      ctx.json({
        results: [
          {
            title: 'Star Wars and the Half-Blood Ewok',
            episode_id: -1,
            opening_crawl:
              "This Ewok's looking a bit peaky. Tensions are in unrest across the stars and also the wars…",
          },
        ],
      })
    )
  })
)

beforeAll(() => server.listen())
afterAll(() => server.close())

describe(StarWars, () => {
  it('display film information returned from API call', async () => {
    // When testing, code that causes React state updates should be wrapped in act(...)
    await act(async () => {
      const component = render(<StarWars />)
      expect(component.baseElement).toMatchSnapshot('StarWars')

      // We wait until what we expect is on the page
      await waitFor(() =>
        component.getByText('Star Wars and the Half-Blood Ewok')
      )
      expect(component.baseElement).toMatchSnapshot('StarWars After Load')

      // By this point, all the text should be rendered
      expect(
        screen.getByText(
          "This Ewok's looking a bit peaky. Tensions are in unrest across the stars and also the wars…"
        )
      ).toBeInTheDocument()
    })
  })
})
