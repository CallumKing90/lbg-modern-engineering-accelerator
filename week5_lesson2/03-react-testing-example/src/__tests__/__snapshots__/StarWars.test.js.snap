// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarWars display film information returned from API call: StarWars 1`] = `
<body>
  <div>
    <main>
      <ul />
    </main>
  </div>
</body>
`;

exports[`StarWars display film information returned from API call: StarWars After Load 1`] = `
<body>
  <div>
    <main>
      <ul>
        <li>
          <h2>
            Star Wars and the Half-Blood Ewok
          </h2>
          <p>
            This Ewok's looking a bit peaky. Tensions are in unrest across the stars and also the warsâ€¦
          </p>
        </li>
      </ul>
    </main>
  </div>
</body>
`;
